builtin/v12/migration/miner.go

```diff
@@ -224,65 +272,111 @@ delta = prevInRoot - inRoot
 
 merge( prevOutRoot, Migrated(delta) )
 */
-func migrateSectorsWithDiff(ctx context.Context, store adt11.Store, inRoot cid.Cid, prevInRoot cid.Cid, prevOutRoot cid.Cid) (cid.Cid, error) {
-	// we have previous work, but the AMT has changed -- diff them
+
+// todo modify hamt based on diff
+func (m minerMigrator) migrateSectorsWithDiff(ctx context.Context, store adt11.Store, minerAddr address.Address, inRoot cid.Cid, prevInRoot cid.Cid, prevOutRoot cid.Cid, prevSectorIndexRoot cid.Cid) (cid.Cid, cid.Cid, error) {
+	// we have previous work
+	// the AMT may or may not have changed -- diff will let us iterate over all the changes
 	diffs, err := amt.Diff(ctx, store, store, prevInRoot, inRoot, amt.UseTreeBitWidth(miner11.SectorsAmtBitwidth))
 	if err != nil {
-		return cid.Undef, xerrors.Errorf("failed to diff old and new Sector AMTs: %w", err)
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to diff old and new Sector AMTs: %w", err)
 	}
 
 	inSectors, err := miner11.LoadSectors(store, inRoot)
 	if err != nil {
-		return cid.Undef, xerrors.Errorf("failed to load inSectors: %w", err)
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to load inSectors: %w", err)
 	}
-
 	prevOutSectors, err := miner12.LoadSectors(store, prevOutRoot)
 	if err != nil {
-		return cid.Undef, xerrors.Errorf("failed to load prevOutSectors: %w", err)
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to load prevOutSectors: %w", err)
+	}
+
+	// load previous HAMT sector index for this specific minerAddr
+	sectorToDealIdHamt, err := builtin.LoadTree(store, prevSectorIndexRoot)
```

Comment:
it's probably safe, but a little wrong to use `LoadTree` here -- that method returns an `ActorTree`, which at least on paper is supposed to be only for "state trees", and not just any HAMT

Reply:
thanks, changed! This was pointed out by zen also 0a40559